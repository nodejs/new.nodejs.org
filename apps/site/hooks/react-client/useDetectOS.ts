'use client';

import { useEffect, useState } from 'react';

import type { UserOS } from '@/types/userOS';
import { detectOS } from '@/util/detectOS';
import { getArchitecture } from '@/util/getArchitecture';
import { getBitness } from '@/util/getBitness';

type UserOSState = {
  os: UserOS;
  bitness: number;
  architecture: string;
};

const useDetectOS = () => {
  const [userOSState, setUserOSState] = useState<UserOSState>({
    os: 'LOADING',
    bitness: 64,
    architecture: '',
  });

  useEffect(() => {
    // If the navigator User Agent indicates a 64-bit OS, we can assume the bitness is 64.
    const uaIndicates64 = /WOW64|Win64|x86_64|x86-64|x64_64|x64;|AMD64/.test(
      navigator.userAgent
    );

    // We immediately set the OS to LOADING, and then we update it with the detected OS.
    // This is due to that initial render set within the state will indicate a mismatch from
    // the server-side rendering versus what the initial state is from the client-side
    setUserOSState(current => ({ ...current, os: detectOS() }));

    // We then update the bitness based on the detected OS and the user agent
    getBitness().then((bitness = '64') =>
      setUserOSState(current => ({
        ...current,
        bitness: bitness === '64' || uaIndicates64 ? 64 : 86,
      }))
    );

    // We then update the architecture based on the detected OS
    getArchitecture().then((architecture = '') =>
      setUserOSState(current => ({ ...current, architecture }))
    );
  }, []);

  return userOSState;
};

export default useDetectOS;
