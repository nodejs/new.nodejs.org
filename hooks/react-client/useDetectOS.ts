'use client';

import { useEffect, useState } from 'react';

import type { UserOS } from '@/types/userOS';
import { detectOS } from '@/util/detectOS';
import { getBitness } from '@/util/getBitness';

type UserOSState = {
  os: UserOS;
  bitness: number;
};

const useDetectOS = () => {
  const [userOSState, setUserOSState] = useState<UserOSState>({
    os: 'OTHER',
    bitness: 86,
  });

  useEffect(() => {
    getBitness().then(bitness => {
      const userAgent = navigator?.userAgent;

      setUserOSState({
        os: detectOS(),
        bitness:
          bitness === '64' ||
          userAgent?.includes('WOW64') ||
          userAgent?.includes('Win64')
            ? 64
            : 86,
      });
    });
  }, []);

  return userOSState;
};

export default useDetectOS;
