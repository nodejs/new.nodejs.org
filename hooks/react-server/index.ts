export { default as useCopyToClipboard } from './useCopyToClipboard';
export { default as useDetectOS } from './useDetectOS';
export { default as useIsCurrentPathname } from './useIsCurrentPathname';
export { default as useMediaQuery } from './useMediaQuery';
export { default as useNotification } from './useNotification';
export { default as useClientContext } from './useClientContext';
export { default as useBlogData } from './useBlogData';
export { default as useSiteNavigation } from './useSiteNavigation';
