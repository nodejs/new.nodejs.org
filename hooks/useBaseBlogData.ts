import { useMemo } from 'react';

import { blogData } from '@/next.json.mjs';

export const useBaseBlogData = (pathname: string) => {
  const { posts, pagination, categories } = blogData;

  const getPostsByCategory = (category: string) =>
    posts.filter(post => post.category === category);

  const getPostsByYear = (year: string) =>
    posts.filter(post => new Date(post.date).getFullYear() === Number(year));

  const getPagination = (currentYear: number | string) => {
    const _currentYear = Number(currentYear);

    return {
      next: pagination.includes(_currentYear + 1)
        ? _currentYear + 1
        : undefined,
      prev: pagination.includes(_currentYear - 1)
        ? _currentYear - 1
        : undefined,
    };
  };

  const currentCategory = useMemo(() => {
    // We split the pathname to retrieve the blog category from it
    // since the URL is usually /{languageCode}/blog/{category}
    // the third path piece is usually the category name
    const [, _pathname, category] = pathname.split('/');

    if (_pathname === 'blog' && category && category.length) {
      return category;
    }

    // if either the pathname does not match to a blog page
    // which should not happen (as this hook should only be used in blog pages)
    // or if there is no category in the URL we return the current year as category name
    // which is always the default category (for example, the blog index)
    return `year-${new Date().getFullYear()}`;
  }, [pathname]);

  return {
    posts,
    categories,
    currentCategory,
    getPostsByCategory,
    getPostsByYear,
    getPagination,
  };
};
