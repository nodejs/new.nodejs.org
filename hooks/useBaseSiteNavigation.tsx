import { useTranslations } from 'next-intl';
import type { RichTranslationValues } from 'next-intl';
import { useMemo } from 'react';

import { siteNavigation } from '@/next.json.mjs';
import type { NavigationEntry, NavigationKeys } from '@/types';

// These are mapped navigation entries. Navigation Entries can have sub-entries
type MappedItems = {
  text: ReturnType<ReturnType<typeof useTranslations>['rich']>;
  link: string;
  key: string;
  level: number;
  items: MappedItems[];
};

// Provides Context replacement for variables within the Link. This is also something that is not going
// to happen in the future with `nodejs/nodejs.dev` codebase
const replaceLinkWithContext = (link: string, context: RichTranslationValues) =>
  Object.entries(context).reduce(
    (finalLink, [find, replace]) =>
      finalLink.replace(
        `{${find}}`,
        typeof replace === 'string' ? replace : ''
      ),
    link
  );

const useBaseSiteNavigation = () => {
  const t = useTranslations();

  const mapNavigationEntries = (
    entries: Record<string, NavigationEntry>,
    context?: Record<string, RichTranslationValues>,
    level = 0
  ): MappedItems[] => {
    const getContext = (key: string) => (context && context[key]) || {};

    const getFormattedMessage = (translationId: string, key: string) =>
      t.rich(translationId, getContext(key));

    return Object.entries(entries).map(([key, item]) => ({
      text: getFormattedMessage(item.translationId, key),
      link: replaceLinkWithContext(item.link, getContext(key)),
      items: item.items
        ? mapNavigationEntries(item.items, context, level + 1)
        : [],
      level,
      key: key,
    }));
  };

  const rootNavigationItems = useMemo(
    () =>
      Object.entries(siteNavigation).reduce(
        (acc, [key, { translationId, link }]) => ({
          ...acc,
          [key]: { translationId, link },
        }),
        {}
      ),
    []
  );

  return {
    navigationItems: mapNavigationEntries(rootNavigationItems),
    getSideNavigation: (
      section: NavigationKeys,
      context?: Record<string, RichTranslationValues>
    ) => {
      const { items, translationId, link } = siteNavigation[section];

      return mapNavigationEntries(
        { [section]: { translationId, link }, ...items },
        context
      );
    },
  };
};

export default useBaseSiteNavigation;
