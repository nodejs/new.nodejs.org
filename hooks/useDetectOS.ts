import { useEffect, useState } from 'react';
import { detectOS } from '../util/detectOS';
import { downloadUrlByOS } from '../util/downloadUrlByOS';

import type { UserOS } from '../types/userOS';

export const useDetectOS = () => {
  const [userOS, setUserOS] = useState<UserOS>('OTHER');
  const [bitness, setBitness] = useState('');

  useEffect(() => {
    setUserOS(detectOS());

    // This is necessary to detect Windows 11 on Edge.
    // [MDN](https://developer.mozilla.org/en-US/docs/Web/API/NavigatorUAData/getHighEntropyValues)
    // [MSFT](https://learn.microsoft.com/en-us/microsoft-edge/web-platform/how-to-detect-win11)
    // @ts-expect-error no types for "userAgentData" because this API is experimental
    if (typeof navigator?.userAgentData?.getHighEntropyValues === 'function') {
      // @ts-expect-error no types for "userAgentData" because this API is experimental
      navigator.userAgentData
        .getHighEntropyValues(['bitness'])
        .then((ua: { bitness: string }) => setBitness(ua.bitness))
        .catch();
    }
  }, []);

  return {
    userOS,
    getDownloadLink: (version: string) =>
      downloadUrlByOS({
        userAgent: navigator?.userAgent,
        userOS,
        version,
        bitness,
      }),
  };
};
