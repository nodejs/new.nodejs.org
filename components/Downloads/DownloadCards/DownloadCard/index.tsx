import classNames from 'classnames';
import { MdGetApp } from 'react-icons/md';
import { Tab } from 'react-tabs';
import Image from 'next/image';
import styles from './index.module.scss';
import { useRouter } from '@/hooks/useRouter';
import type { FC } from 'react';
import type { UserOS } from '@/types/userOS';

type DownloadCardProps = {
  os: UserOS;
  icon: string;
  label: string;
  download: string;
  filename: string;
  selected: boolean;
  onSelect: (os: string) => void;
};

const DownloadCard: FC<DownloadCardProps> = ({
  os,
  icon,
  label,
  download,
  filename,
  selected,
  onSelect,
}) => {
  const { basePath } = useRouter();

  const tabClassNames = classNames(styles.downloadCard, {
    [styles.downloadCardActive]: selected,
  });
  const selectCard = () => onSelect(os);
  const size = selected ? 56 : 48;

  return (
    <Tab
      className={tabClassNames}
      id={`download-card-${os}`}
      key={os}
      onClick={selectCard}
      selectedClassName={styles.downloadCardActive}
      tabIndex="0"
    >
      <div className={styles.top}>
        <Image
          height={size}
          width={size}
          src={`${basePath}/static/images/logos/${icon}`}
          alt={label}
        />
        {selected && (
          <a className={styles.link} href={download}>
            <MdGetApp />
          </a>
        )}
      </div>
      <p className={styles.label}>{label}</p>
      <p className={styles.filename} data-testid="lts">
        {filename}
      </p>
    </Tab>
  );
};

export default DownloadCard;
