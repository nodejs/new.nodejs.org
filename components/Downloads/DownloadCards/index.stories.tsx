import { createMock } from 'storybook-addon-module-mock';
import * as hooks from '@/hooks/useDetectOS';
import DownloadCards from './index';
import type { Meta as MetaObj, StoryObj } from '@storybook/react';

type Story = StoryObj<typeof DownloadCards>;
type Meta = MetaObj<typeof DownloadCards>;

export const Default: Story = {
  args: {
    versionWithPrefix: 'v18.15.0',
  },
  parameters: {
    moduleMock: {
      mock: () => {
        const mock = createMock(hooks, 'useDetectOS');
        mock.mockReturnValue({
          os: 'WIN',
          bitness: 64,
        });
      },
    },
  },
};

export default { component: DownloadCards } as Meta;
