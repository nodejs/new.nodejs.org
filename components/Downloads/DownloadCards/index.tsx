import { Tabs, TabList } from 'react-tabs';
import { useState, useEffect } from 'react';
import styles from './index.module.scss';
import DownloadCard from './DownloadCard/index';
import type { FC, KeyboardEvent } from 'react';
import type { UserOS } from '../../../types/userOS';

type DownloadCardsProps = {
  version: string;
  userOS: UserOS;
};

const DownloadCards: FC<DownloadCardsProps> = ({ version, userOS }) => {
  const [selected, setSelected] = useState('');

  useEffect(() => {
    setSelected(!['WIN', 'MAC'].includes(userOS) ? 'SOURCECODE' : userOS);
  }, [userOS]);

  const downloadTypes = [
    {
      name: 'WIN',
      icon: 'windows-download-logo.svg',
      label: 'Windows Installer',
      download: `https://nodejs.org/dist/${version}/node-${version}-x86.msi`,
      filename: `node-${version}-x86.msi`,
    },
    {
      name: 'MAC',
      icon: 'mac-download-logo.svg',
      label: 'MAC Installer',
      download: `https://nodejs.org/dist/${version}/node-${version}.pkg`,
      filename: `node-${version}.pkg`,
    },
    {
      name: 'SOURCECODE',
      icon: 'source-code-download-logo.svg',
      label: 'Source Code',
      download: `https://nodejs.org/dist/${version}/node-${version}.tar.gz`,
      filename: `node-${version}.tar.gz`,
    },
  ];

  return (
    <Tabs tabIndex={-1} className={styles.downloadCardsWrapper}>
      <TabList
        className={styles.downloadCards}
        role="tablist"
        onKeyDown={(e: KeyboardEvent) => {
          const current = downloadTypes.findIndex(d => d.name === selected);

          let step = 0;
          if (e.key === 'ArrowLeft') {
            step = -1;
          } else if (e.key === 'ArrowRight') {
            step = 1;
          }
          if (!step) return;

          const next =
            (current + step + downloadTypes.length) % downloadTypes.length;
          setSelected(downloadTypes[next].name);
        }}
      >
        {downloadTypes.map(d => (
          <DownloadCard
            key={d.name}
            name={d.name}
            icon={d.icon}
            label={d.label}
            download={d.download}
            filename={d.filename}
            selected={selected === d.name}
            onSelect={setSelected}
          />
        ))}
      </TabList>
    </Tabs>
  );
};

export default DownloadCards;
