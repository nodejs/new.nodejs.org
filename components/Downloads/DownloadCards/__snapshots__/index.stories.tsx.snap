// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Downloads/DownloadCards Default smoke-test 1`] = `
<div tabindex="-1"
     data-rttabs="true"
>
  <ul role="tablist">
    <li role="tab"
        id="tabdownload-card-WIN"
        aria-selected="false"
        aria-disabled="false"
        aria-controls="paneldownload-card-WIN"
        tabindex="0"
        data-rttab="true"
    >
      <div>
        <img alt="Windows Installer"
             loading="lazy"
             width="56"
             height="56"
             decoding="async"
             data-nimg="1"
             srcset="/static/images/logos/windows-download-logo.svg?w=64&amp;q=75 1x, /static/images/logos/windows-download-logo.svg?w=128&amp;q=75 2x"
             src="/static/images/logos/windows-download-logo.svg?w=128&amp;q=75"
             style="color: transparent;"
        >
        <a href="https://nodejs.org/dist/v18.15.0/node-v18.15.0-x64.msi">
          <svg stroke="currentColor"
               fill="currentColor"
               stroke-width="0"
               viewbox="0 0 24 24"
               height="1em"
               width="1em"
               xmlns="http://www.w3.org/2000/svg"
          >
            <path fill="none"
                  d="M0 0h24v24H0z"
            >
            </path>
            <path d="M19 9h-4V3H9v6H5l7 7 7-7zM5 18v2h14v-2H5z">
            </path>
          </svg>
        </a>
      </div>
      <p>
        Windows Installer
      </p>
      <p data-testid="lts">
        node-v18.15.0-x64.msi
      </p>
    </li>
    <li role="tab"
        id="tabdownload-card-MAC"
        aria-selected="false"
        aria-disabled="false"
        aria-controls="paneldownload-card-MAC"
        tabindex="0"
        data-rttab="true"
    >
      <div>
        <img alt="MAC Installer"
             loading="lazy"
             width="48"
             height="48"
             decoding="async"
             data-nimg="1"
             srcset="/static/images/logos/mac-download-logo.svg?w=48&amp;q=75 1x, /static/images/logos/mac-download-logo.svg?w=96&amp;q=75 2x"
             src="/static/images/logos/mac-download-logo.svg?w=96&amp;q=75"
             style="color: transparent;"
        >
      </div>
      <p>
        MAC Installer
      </p>
      <p data-testid="lts">
        node-v18.15.0.pkg
      </p>
    </li>
    <li role="tab"
        id="tabdownload-card-OTHER"
        aria-selected="false"
        aria-disabled="false"
        aria-controls="paneldownload-card-OTHER"
        tabindex="0"
        data-rttab="true"
    >
      <div>
        <img alt="Source Code"
             loading="lazy"
             width="48"
             height="48"
             decoding="async"
             data-nimg="1"
             srcset="/static/images/logos/source-code-download-logo.svg?w=48&amp;q=75 1x, /static/images/logos/source-code-download-logo.svg?w=96&amp;q=75 2x"
             src="/static/images/logos/source-code-download-logo.svg?w=96&amp;q=75"
             style="color: transparent;"
        >
      </div>
      <p>
        Source Code
      </p>
      <p data-testid="lts">
        node-v18.15.0.tar.gz
      </p>
    </li>
  </ul>
</div>
`;
